### DISCUSSION TITLE
Delete from list of remote items

### DISCUSSION THREAD

**Discussion by tylerjames** [pointfreeco/swift-composable-architecture#1306] [Q&A]
Let's say I have struct called `Friend` and a list of `Friend`s that is fetched remotely from a server.

I have a list view (UITableView) that shows the first and last name of each `Friend`. Tapping on an item goes to the edit view for that `Friend` which shows many more details and actions that can be done and it is represented by `EditFriendState`.

So the list view is driven by `IdentifiedArrayOf<EditFriendState>`. The list view allows a swipe to delete action on any friend which will remove the friend from the list immediately and send a remote request to delete the friend from the back end.

**The problem is that this delete request could fail**, in which case I'd need to restore the friend to the list, so I can't really remove it from my `IdentifiedArrayOf<EditFriendState>` until after the delete request is successful. 

First I was thinking I could add an `isPendingDelete` to the `EditFriendState` but that seems wrong because the pending delete state seems to be more of a property of the list view rather than the Edit Friend view that's represented by `EditFriendState`

I figured another alternative would be to add a new property to my `FriendListState` that is an array of `Friend` IDs that are currently pending delete `friendsPendingDelete`. And then I would filter my table view datasource to exclude any friends in the pending list and then:

- if delete succeeds remove the Friend from the  Friend list and remove their ID from the `friendsPendingDelete` 
- if delete fails then just remove their ID from the `friendsPendingDelete` list which will restore them to the Friend list.

Does that make sense or is there a canonical way of handling situations like this?

**Comment by DivineDominion** [comment]
Sorry for resurrecting, but it's an interesting problem.

- The user's intent can be expressed in local state (swipe to delete)
- Deletion request (to the server) can fail

This is an interesting question that comes up during intermediate stages of development -- apart from, say, social media apps, where you only have a dumb client:

1. Start with in-memory storage to try out everything
2. Move to live storage on a remote server (⬅️ OP was here)
3. Implement offline support and sync

Once you move to step 3, the problem of representing pending states goes away, because the app will represent local changes and sync with the server _eventually_. Sync may fail, but the local source of truth after swiping to delete is that the `Friend` has been deleted. It stays gone, no matter the network connection. If the server then ingests the changes and decides to resurrect or otherwise not delete the `Friend`, your after-sync state will make the friend pop up again.

If there's never going to be a step 3, though, a request/pending/response triple (like for every web request, really) is probably the way to go!

- Blocking, on the action level: e.g. show spinner, wait for response, then perform deletion
- Non-blocking, in the state (probably `FriendListState` from the example): change the friend list item from a 'just show the friend' state to 'processing changes/awaiting server response' state. Nesting `enum`s will work fine here.
    - You can denote pending edits separately from pending deletions, so that the UI can proactively skip the list item in the UI. Swipe-to-delete then properly removed the visual representation of the list item instead of showing a row with a spinner.
